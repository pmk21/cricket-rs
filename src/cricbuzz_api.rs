use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzOverSeparator {
    pub score: u32,
    pub wickets: u32,
    pub innings_id: u32,
    pub o_summary: String,
    pub runs: u32,
    pub bat_striker_ids: Vec<u32>,
    pub bat_striker_names: Vec<String>,
    pub bat_striker_runs: u32,
    pub bat_striker_balls: u32,
    pub bat_non_striker_ids: Vec<u32>,
    pub bat_non_striker_names: Vec<String>,
    pub bat_non_striker_runs: u32,
    pub bat_non_striker_balls: u32,
    pub bowl_ids: Vec<u32>,
    pub bowl_names: Vec<String>,
    pub bowl_overs: f32,
    pub bowl_maidens: u32,
    pub bowl_runs: u32,
    pub bowl_wickets: u32,
    pub timestamp: u64,
    pub over_num: f32,
    pub bat_team_name: String,
    pub event: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzBatsmanStriker {
    pub bat_balls: u32,
    pub bat_dots: u32,
    pub bat_fours: u32,
    pub bat_id: u32,
    pub bat_name: String,
    pub bat_mins: u32,
    pub bat_runs: u32,
    pub bat_sixes: u32,
    pub bat_strike_rate: f32,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzBowlerStriker {
    pub bowl_id: u32,
    pub bowl_name: String,
    pub bowl_maidens: u32,
    pub bowl_noballs: u32,
    pub bowl_ovs: f32,
    pub bowl_runs: u32,
    pub bowl_wides: u32,
    pub bowl_wkts: u32,
    pub bowl_econ: f32,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzCommentary {
    pub comm_text: String,
    pub timestamp: u64,
    pub ball_nbr: u32,
    pub over_number: f32,
    pub innings_id: u32,
    pub event: String,
    pub bat_team_name: String,
    pub commentary_formats: Vec<String>, // Not sure about data type
    pub over_separator: CricbuzzOverSeparator,
    pub batsman_striker: CricbuzzBatsmanStriker,
    pub bowler_striker: CricbuzzBowlerStriker,
}

#[derive(Debug, Deserialize)]
pub struct CricbuzzMatchHeaderTossResults {
    pub toss_winner_id: u32,
    pub toss_winner_name: String,
    pub decision: String,
}

#[derive(Debug, Deserialize)]
pub struct CricbuzzMatchHeaderResults {
    pub winning_team: String,
    pub win_by_runs: bool,
    pub win_by_innings: bool,
}

#[derive(Debug, Deserialize)]
pub struct CricbuzzMatchHeaderRevisedTarget {
    pub reason: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMatchHeaderMatchTeamInfo {
    pub batting_team_id: u32,
    pub batting_team_short_name: String,
    pub bowling_team_id: u32,
    pub bowling_team_short_name: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMatchHeaderTeam {
    pub id: u32,
    pub name: String,
    pub player_details: Vec<String>,
    pub short_name: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMatchHeader {
    pub match_id: u32,
    pub match_description: String,
    pub match_format: String,
    pub match_type: String,
    pub complete: bool,
    pub domestic: bool,
    pub match_start_timestamp: u64,
    pub match_complete_timestamp: u64,
    pub day_night: bool,
    pub year: u32,
    pub day_number: u32,
    pub state: String,
    pub status: String,
    pub toss_results: CricbuzzMatchHeaderTossResults,
    pub result: CricbuzzMatchHeaderResults,
    pub revised_target: CricbuzzMatchHeaderRevisedTarget,
    pub players_of_the_match: Vec<String>,
    pub players_of_the_series: Vec<String>,
    pub match_team_info: Vec<CricbuzzMatchHeaderMatchTeamInfo>,
    pub is_match_not_covered: bool,
    pub team1: CricbuzzMatchHeaderTeam,
    pub team2: CricbuzzMatchHeaderTeam,
    pub series_desc: String,
    pub series_id: u32,
    pub series_name: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMiniscoreBatsman {
    pub bat_balls: u32,
    pub bat_dots: u32,
    pub bat_fours: u32,
    pub bat_id: u32,
    pub bat_name: String,
    pub bat_mins: u32,
    pub bat_sixes: u32,
    pub bat_strike_rate: f32,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMiniscoreBatTeam {
    pub team_id: u32,
    pub team_score: u32,
    pub team_wkts: u32,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMiniscoreBowler {
    pub bowl_id: u32,
    pub bowl_name: String,
    pub bowl_maidens: u32,
    pub bowl_noballs: u32,
    pub bowl_ovs: f32,
    pub bowl_runs: u32,
    pub bowl_wides: u32,
    pub bowl_wkts: u32,
    pub bowl_econ: f32,
}

#[derive(Debug, Deserialize)]
pub struct CricbuzzMiniscorePartnership {
    pub balls: u32,
    pub runs: u32,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMiniscoreMatchScoreDetailsInningsScore {
    pub innings_id: u32,
    pub bat_team_id: u32,
    pub bat_team_name: String,
    pub score: u32,
    pub wickets: u32,
    pub overs: f32,
    pub is_declared: bool,
    pub is_follow_on: bool,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMiniscoreMatchScoreDetailsTossResults {
    pub toss_winner_id: u32,
    pub toss_winner_name: String,
    pub decision: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMiniscoreMatchScoreDetailsMatchTeamInfo {
    pub batting_team_id: u32,
    pub batting_team_short_name: String,
    pub bowling_team_id: u32,
    pub bowling_team_short_name: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMiniscoreMatchScoreDetails {
    pub match_id: u32,
    pub innings_score_list: Vec<CricbuzzMiniscoreMatchScoreDetailsInningsScore>,
    pub toss_results: CricbuzzMiniscoreMatchScoreDetailsTossResults,
    pub match_team_info: Vec<CricbuzzMiniscoreMatchScoreDetailsMatchTeamInfo>,
    pub is_match_not_covered: bool,
    pub match_format: String,
    pub state: String,
    pub custom_status: String,
    pub highlighted_team_id: u32,
}

#[derive(Debug, Deserialize)]
pub struct CricbuzzMiniscoreLatestPerformance {
    pub runs: u32,
    pub wkts: u32,
    pub label: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMiniscoreMatchUdrs {
    pub match_id: u32,
    pub innings_id: u32,
    pub timestamp: String,
    pub team1_id: u32,
    pub team1_remaining: u32,
    pub team1_successful: u32,
    pub team1_unsuccessful: u32,
    pub team2_id: u32,
    pub team2_remaining: u32,
    pub team2_successful: u32,
    pub team2_unsuccessful: u32,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzMiniscore {
    pub innings_id: u32,
    pub batsman_striker: CricbuzzMiniscoreBatsman,
    pub batsman_non_striker: CricbuzzMiniscoreBatsman,
    pub bat_team: CricbuzzMiniscoreBatTeam,
    pub bowler_striker: CricbuzzMiniscoreBowler,
    pub bowler_non_striker: CricbuzzMiniscoreBowler,
    pub overs: f32,
    pub recent_ovs_stats: String,
    pub partner_ship: CricbuzzMiniscorePartnership,
    pub current_run_rate: f32,
    pub required_run_rate: f32,
    // TODO: Value is not always present
    // pub last_wicket: String,
    pub match_score_details: CricbuzzMiniscoreMatchScoreDetails,
    pub latest_performance: Vec<CricbuzzMiniscoreLatestPerformance>,
    // ppData: Not parsed
    // TODO: Value is not always present
    // pub match_udrs: CricbuzzMiniscoreMatchUdrs,
    // overSummaryList: Not parsed
    // TODO: Value is not always present
    // pub overs_rem: f32,
    pub status: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CricbuzzJson {
    // pub commentary_list: Vec<CricbuzzCommentary>,
    // pub match_header: CricbuzzMatchHeader,
    pub miniscore: CricbuzzMiniscore,
    pub page: String,
    pub enable_no_content: bool,
}
