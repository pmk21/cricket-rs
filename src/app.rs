use select::predicate::{Class, Name, Predicate};
use select::{document::Document, node::Node, predicate::Attr};

use crate::cricbuzz_api::*;

#[derive(Debug, Default, Clone)]
pub struct BowlerInfo {
    pub name: String,
    pub overs: String,
    pub maidens: String,
    pub runs: String,
    pub wickets: String,
    pub no_balls: String,
    pub wides: String,
    pub economy: String,
}

#[derive(Default, Debug, Clone)]
pub struct BatsmanInfo {
    pub name: String,
    pub status: String,
    pub runs: String,
    pub balls: String,
    pub fours: String,
    pub sixes: String,
    pub strike_rate: String,
}

#[derive(Debug, Default, Clone)]
pub struct MatchInningsInfo {
    pub batsman_details: Vec<BatsmanInfo>,
    pub yet_to_bat: String,
    pub bowler_details: Vec<BowlerInfo>,
}

pub struct MatchInfo {
    pub match_short_name: String,
    pub cricbuzz_match_id: u32,
    pub cricbuzz_match_api_link: String,
    pub cricbuzz_info: CricbuzzJson,
    pub scorecard: Vec<MatchInningsInfo>,
    pub scorecard_scroll: u16,
}

pub struct App {
    pub matches_info: Vec<MatchInfo>,
    pub focused_tab: u32,
}

const CRICBUZZ_URL: &str = "https://www.cricbuzz.com";
const CRICBUZZ_MATCH_API: &str = "https://www.cricbuzz.com/api/cricket-match/commentary/";
const CRICBUZZ_MATCH_SCORECARD_API: &str = "https://www.cricbuzz.com/api/html/cricket-scorecard/";

impl App {
    pub async fn new() -> App {
        let mut matches_info = vec![];
        let mut match_name_id: Vec<(String, String)> = vec![];
        let mut scorecard: Vec<MatchInningsInfo> = vec![];

        match get_all_live_matches_id_and_short_name().await {
            Ok(v) => match_name_id = v,
            Err(e) => {
                println!("{:?}", e);
            }
        };

        for (name, id) in &match_name_id {
            let match_id: u32 = id.parse().unwrap();
            if let Ok(json) = get_match_info_from_id(match_id).await {
                if let Ok(_) = prepare_scorecard(match_id, &mut scorecard).await {
                    matches_info.push(MatchInfo::new(
                        name.to_string(),
                        match_id,
                        format!("{}{}", String::from(CRICBUZZ_MATCH_API), match_id),
                        json,
                        scorecard.clone(),
                    ));
                    scorecard.clear();
                } else {
                    matches_info.push(MatchInfo::new(
                        name.to_string(),
                        match_id,
                        format!("{}{}", String::from(CRICBUZZ_MATCH_API), match_id),
                        json,
                        vec![],
                    ));
                }
            }
        }

        let focused_tab = 0;

        App {
            matches_info,
            focused_tab,
        }
    }
}

impl MatchInfo {
    fn new(
        match_short_name: String,
        cricbuzz_match_id: u32,
        cricbuzz_match_api_link: String,
        cricbuzz_info: CricbuzzJson,
        scorecard: Vec<MatchInningsInfo>,
    ) -> MatchInfo {
        MatchInfo {
            match_short_name,
            cricbuzz_match_id,
            cricbuzz_match_api_link,
            cricbuzz_info,
            scorecard,
            scorecard_scroll: 0,
        }
    }
}

// TODO: Need to improve method of getting all matches
async fn get_all_live_matches_id_and_short_name(
) -> Result<Vec<(String, String)>, Box<dyn std::error::Error>> {
    let resp_html = reqwest::get(CRICBUZZ_URL).await?.text().await?;

    let document = Document::from(resp_html.as_str());
    let mut match_id_name = vec![];

    for node in document.find(Class("cb-mat-mnu").descendant(Name("a"))) {
        // This check might break in the future
        if !node.text().is_empty() && !node.text().eq("MATCHES") {
            if node.text().split('-').nth(1).unwrap().trim().eq("Live") {
                match node.attr("href") {
                    Some(link) => {
                        let split_str: Vec<&str> = link.split('/').collect();
                        match_id_name.push((node.text(), String::from(split_str[2])));
                    }
                    _ => {}
                }
            }
        }
    }
    Ok(match_id_name)
}

async fn get_match_info_from_id(match_id: u32) -> Result<CricbuzzJson, Box<dyn std::error::Error>> {
    let resp = reqwest::get(format!("{}{}", String::from(CRICBUZZ_MATCH_API), match_id))
        .await?
        .text()
        .await?;

    let res: CricbuzzJson = serde_json::from_str(&resp).unwrap();
    Ok(res)
}

async fn prepare_scorecard(
    id: u32,
    scorecard: &mut Vec<MatchInningsInfo>,
) -> Result<(), Box<dyn std::error::Error>> {
    let resp_html = reqwest::get(format!(
        "{}{}",
        String::from(CRICBUZZ_MATCH_SCORECARD_API),
        id
    ))
    .await?
    .text()
    .await?;

    let document = Document::from(resp_html.as_str());

    for i in 1..5 {
        if let Some(inngs) = document
            .find(Attr("id", format!("innings_{}", i.to_string()).as_ref()))
            .nth(0)
        {
            populate_innings_info(&inngs, scorecard);
        }
    }

    Ok(())
}

fn populate_innings_info(inngs: &Node, scorecard: &mut Vec<MatchInningsInfo>) {
    // WARNING: This scorecard parsing function might break easily!

    let mut match_inngs_info = MatchInningsInfo::default();

    let mut count = 0;

    // TODO: Split this into functions
    // Prints each batsman's scorecard info
    let bat_info = inngs.children().nth(1).unwrap();
    let mut batsman_info = BatsmanInfo::default();
    for node in bat_info.children().skip(4) {
        if node.attr("class").is_some() {
            if node.children().count() == 15 {
                for inner_node in node.children() {
                    if !inner_node.text().trim().is_empty() {
                        if count == 0 {
                            batsman_info.name = inner_node.text().trim().to_string();
                        } else if count == 1 {
                            batsman_info.status = inner_node.text().trim().to_string();
                        } else if count == 2 {
                            batsman_info.runs = inner_node.text().trim().to_string();
                        } else if count == 3 {
                            batsman_info.balls = inner_node.text().trim().to_string();
                        } else if count == 4 {
                            batsman_info.fours = inner_node.text().trim().to_string();
                        } else if count == 5 {
                            batsman_info.sixes = inner_node.text().trim().to_string();
                        } else if count == 6 {
                            batsman_info.strike_rate = inner_node.text().trim().to_string();
                        }

                        count += 1;
                    }
                }
                count = 0;
                match_inngs_info.batsman_details.push(batsman_info.clone());
            }
        }
    }

    // Fall of Wickets
    // let fow_hdr = inngs.children().nth(3).unwrap();
    // println!("{}", fow_hdr.text());

    // Shows all wickets fallen
    // let fow_all = inngs.children().nth(4).unwrap();
    // println!("{}", fow_all.text());

    // TODO: Split this into functions
    // Bowler Scorecard Info - Bowler O M R W NB WD ECO
    count = 0;
    if let Some(bowl_info) = inngs.children().nth(6) {
        let mut bowler_info = BowlerInfo::default();
        for n in bowl_info.children().skip(2) {
            if n.attr("class").is_some() {
                for n1 in n.children() {
                    if !n1.text().trim().is_empty() {
                        if count == 0 {
                            bowler_info.name = n1.text().trim().to_string();
                        } else if count == 1 {
                            bowler_info.overs = n1.text().trim().to_string();
                        } else if count == 2 {
                            bowler_info.maidens = n1.text().trim().to_string();
                        } else if count == 3 {
                            bowler_info.runs = n1.text().trim().to_string();
                        } else if count == 4 {
                            bowler_info.wickets = n1.text().trim().to_string();
                        } else if count == 5 {
                            bowler_info.no_balls = n1.text().trim().to_string();
                        } else if count == 6 {
                            bowler_info.wides = n1.text().trim().to_string();
                        } else if count == 7 {
                            bowler_info.economy = n1.text().trim().to_string();
                        }

                        count += 1;
                    }
                }
                match_inngs_info.bowler_details.push(bowler_info.clone());
                count = 0;
            }
        }
    }

    scorecard.push(match_inngs_info);
}
