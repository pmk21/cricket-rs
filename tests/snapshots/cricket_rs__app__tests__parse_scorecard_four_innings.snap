---
source: src/app.rs
expression: scorecard

---
[
    MatchInningsInfo {
        batsman_details: [
            BatsmanInfo {
                name: "Zak Crawley",
                status: "lbw b Axar",
                runs: "53",
                balls: "84",
                fours: "10",
                sixes: "0",
                strike_rate: "63.10",
            },
            BatsmanInfo {
                name: "D Sibley",
                status: "c Rohit b Ishant",
                runs: "0",
                balls: "7",
                fours: "0",
                sixes: "0",
                strike_rate: "0.00",
            },
            BatsmanInfo {
                name: "Bairstow",
                status: "lbw b Axar",
                runs: "0",
                balls: "9",
                fours: "0",
                sixes: "0",
                strike_rate: "0.00",
            },
            BatsmanInfo {
                name: "Root (c)",
                status: "lbw b Ashwin",
                runs: "17",
                balls: "37",
                fours: "2",
                sixes: "0",
                strike_rate: "45.95",
            },
            BatsmanInfo {
                name: "Stokes",
                status: "lbw b Axar",
                runs: "6",
                balls: "24",
                fours: "1",
                sixes: "0",
                strike_rate: "25.00",
            },
            BatsmanInfo {
                name: "Ollie Pope",
                status: "b Ashwin",
                runs: "1",
                balls: "12",
                fours: "0",
                sixes: "0",
                strike_rate: "8.33",
            },
            BatsmanInfo {
                name: "Foakes (wk)",
                status: "b Axar",
                runs: "12",
                balls: "58",
                fours: "1",
                sixes: "0",
                strike_rate: "20.69",
            },
            BatsmanInfo {
                name: "Jofra Archer",
                status: "b Axar",
                runs: "11",
                balls: "18",
                fours: "2",
                sixes: "0",
                strike_rate: "61.11",
            },
            BatsmanInfo {
                name: "Jack Leach",
                status: "c Pujara b Ashwin",
                runs: "3",
                balls: "14",
                fours: "0",
                sixes: "0",
                strike_rate: "21.43",
            },
            BatsmanInfo {
                name: "Broad",
                status: "c Bumrah b Axar",
                runs: "3",
                balls: "29",
                fours: "0",
                sixes: "0",
                strike_rate: "10.34",
            },
            BatsmanInfo {
                name: "James Anderson",
                status: "not out",
                runs: "0",
                balls: "3",
                fours: "0",
                sixes: "0",
                strike_rate: "0.00",
            },
        ],
        yet_to_bat: "",
        bowler_details: [
            BowlerInfo {
                name: "Ishant",
                overs: "5",
                maidens: "1",
                runs: "26",
                wickets: "1",
                no_balls: "2",
                wides: "0",
                economy: "5.20",
            },
            BowlerInfo {
                name: "Bumrah",
                overs: "6",
                maidens: "3",
                runs: "19",
                wickets: "0",
                no_balls: "1",
                wides: "0",
                economy: "3.20",
            },
            BowlerInfo {
                name: "Axar",
                overs: "21.4",
                maidens: "6",
                runs: "38",
                wickets: "6",
                no_balls: "0",
                wides: "0",
                economy: "1.80",
            },
            BowlerInfo {
                name: "Ashwin",
                overs: "16",
                maidens: "6",
                runs: "26",
                wickets: "3",
                no_balls: "0",
                wides: "0",
                economy: "1.60",
            },
        ],
    },
    MatchInningsInfo {
        batsman_details: [
            BatsmanInfo {
                name: "Rohit",
                status: "lbw b Jack Leach",
                runs: "66",
                balls: "96",
                fours: "11",
                sixes: "0",
                strike_rate: "68.75",
            },
            BatsmanInfo {
                name: "Shubman Gill",
                status: "c Zak Crawley b Jofra Archer",
                runs: "11",
                balls: "51",
                fours: "2",
                sixes: "0",
                strike_rate: "21.57",
            },
            BatsmanInfo {
                name: "Pujara",
                status: "lbw b Jack Leach",
                runs: "0",
                balls: "4",
                fours: "0",
                sixes: "0",
                strike_rate: "0.00",
            },
            BatsmanInfo {
                name: "Kohli (c)",
                status: "b Jack Leach",
                runs: "27",
                balls: "58",
                fours: "3",
                sixes: "0",
                strike_rate: "46.55",
            },
            BatsmanInfo {
                name: "Rahane",
                status: "lbw b Jack Leach",
                runs: "7",
                balls: "25",
                fours: "1",
                sixes: "0",
                strike_rate: "28.00",
            },
            BatsmanInfo {
                name: "Pant (wk)",
                status: "c Foakes b Root",
                runs: "1",
                balls: "8",
                fours: "0",
                sixes: "0",
                strike_rate: "12.50",
            },
            BatsmanInfo {
                name: "Ashwin",
                status: "c Zak Crawley b Root",
                runs: "17",
                balls: "32",
                fours: "3",
                sixes: "0",
                strike_rate: "53.12",
            },
            BatsmanInfo {
                name: "Washington Sundar",
                status: "b Root",
                runs: "0",
                balls: "12",
                fours: "0",
                sixes: "0",
                strike_rate: "0.00",
            },
            BatsmanInfo {
                name: "Axar",
                status: "c D Sibley b Root",
                runs: "0",
                balls: "2",
                fours: "0",
                sixes: "0",
                strike_rate: "0.00",
            },
            BatsmanInfo {
                name: "Ishant",
                status: "not out",
                runs: "10",
                balls: "20",
                fours: "0",
                sixes: "1",
                strike_rate: "50.00",
            },
            BatsmanInfo {
                name: "Bumrah",
                status: "lbw b Root",
                runs: "1",
                balls: "12",
                fours: "0",
                sixes: "0",
                strike_rate: "8.33",
            },
        ],
        yet_to_bat: "",
        bowler_details: [
            BowlerInfo {
                name: "James Anderson",
                overs: "13",
                maidens: "8",
                runs: "20",
                wickets: "0",
                no_balls: "0",
                wides: "0",
                economy: "1.50",
            },
            BowlerInfo {
                name: "Broad",
                overs: "6",
                maidens: "1",
                runs: "16",
                wickets: "0",
                no_balls: "0",
                wides: "0",
                economy: "2.70",
            },
            BowlerInfo {
                name: "Jofra Archer",
                overs: "5",
                maidens: "2",
                runs: "24",
                wickets: "1",
                no_balls: "0",
                wides: "0",
                economy: "4.80",
            },
            BowlerInfo {
                name: "Jack Leach",
                overs: "20",
                maidens: "2",
                runs: "54",
                wickets: "4",
                no_balls: "0",
                wides: "0",
                economy: "2.70",
            },
            BowlerInfo {
                name: "Stokes",
                overs: "3",
                maidens: "0",
                runs: "19",
                wickets: "0",
                no_balls: "0",
                wides: "1",
                economy: "6.30",
            },
            BowlerInfo {
                name: "Root (c)",
                overs: "6.2",
                maidens: "3",
                runs: "8",
                wickets: "5",
                no_balls: "0",
                wides: "0",
                economy: "1.30",
            },
        ],
    },
    MatchInningsInfo {
        batsman_details: [
            BatsmanInfo {
                name: "Zak Crawley",
                status: "b Axar",
                runs: "0",
                balls: "1",
                fours: "0",
                sixes: "0",
                strike_rate: "0.00",
            },
            BatsmanInfo {
                name: "D Sibley",
                status: "c Pant b Axar",
                runs: "7",
                balls: "25",
                fours: "0",
                sixes: "0",
                strike_rate: "28.00",
            },
            BatsmanInfo {
                name: "Bairstow",
                status: "b Axar",
                runs: "0",
                balls: "2",
                fours: "0",
                sixes: "0",
                strike_rate: "0.00",
            },
            BatsmanInfo {
                name: "Root (c)",
                status: "lbw b Axar",
                runs: "19",
                balls: "45",
                fours: "0",
                sixes: "0",
                strike_rate: "42.22",
            },
            BatsmanInfo {
                name: "Stokes",
                status: "lbw b Ashwin",
                runs: "25",
                balls: "34",
                fours: "3",
                sixes: "0",
                strike_rate: "73.53",
            },
            BatsmanInfo {
                name: "Ollie Pope",
                status: "b Ashwin",
                runs: "12",
                balls: "15",
                fours: "2",
                sixes: "0",
                strike_rate: "80.00",
            },
            BatsmanInfo {
                name: "Foakes (wk)",
                status: "lbw b Axar",
                runs: "8",
                balls: "28",
                fours: "0",
                sixes: "0",
                strike_rate: "28.57",
            },
            BatsmanInfo {
                name: "Jofra Archer",
                status: "lbw b Ashwin",
                runs: "0",
                balls: "2",
                fours: "0",
                sixes: "0",
                strike_rate: "0.00",
            },
            BatsmanInfo {
                name: "Jack Leach",
                status: "c Rahane b Ashwin",
                runs: "9",
                balls: "22",
                fours: "0",
                sixes: "1",
                strike_rate: "40.91",
            },
            BatsmanInfo {
                name: "Broad",
                status: "not out",
                runs: "1",
                balls: "7",
                fours: "0",
                sixes: "0",
                strike_rate: "14.29",
            },
            BatsmanInfo {
                name: "James Anderson",
                status: "c Pant b Washington Sundar",
                runs: "0",
                balls: "3",
                fours: "0",
                sixes: "0",
                strike_rate: "0.00",
            },
        ],
        yet_to_bat: "",
        bowler_details: [
            BowlerInfo {
                name: "Axar",
                overs: "15",
                maidens: "0",
                runs: "32",
                wickets: "5",
                no_balls: "0",
                wides: "0",
                economy: "2.10",
            },
            BowlerInfo {
                name: "Ashwin",
                overs: "15",
                maidens: "3",
                runs: "48",
                wickets: "4",
                no_balls: "0",
                wides: "0",
                economy: "3.20",
            },
            BowlerInfo {
                name: "Washington Sundar",
                overs: "0.4",
                maidens: "0",
                runs: "1",
                wickets: "1",
                no_balls: "0",
                wides: "0",
                economy: "1.50",
            },
        ],
    },
    MatchInningsInfo {
        batsman_details: [
            BatsmanInfo {
                name: "Rohit Sharma",
                status: "not out",
                runs: "25",
                balls: "25",
                fours: "3",
                sixes: "1",
                strike_rate: "100.00",
            },
            BatsmanInfo {
                name: "Shubman Gill",
                status: "not out",
                runs: "15",
                balls: "21",
                fours: "1",
                sixes: "1",
                strike_rate: "71.43",
            },
        ],
        yet_to_bat: "",
        bowler_details: [
            BowlerInfo {
                name: "Jack Leach",
                overs: "4",
                maidens: "1",
                runs: "15",
                wickets: "0",
                no_balls: "0",
                wides: "0",
                economy: "3.80",
            },
            BowlerInfo {
                name: "Joe Root (c)",
                overs: "3.4",
                maidens: "0",
                runs: "25",
                wickets: "0",
                no_balls: "0",
                wides: "0",
                economy: "6.80",
            },
        ],
    },
]
